### PR TITLE
[DLLM]Fast-dLLM-v2 support with HierarchyBlock algorithm for parallel decoding

### PR DESCRIPTION
**Pull request by Monstertail**
## Motivation

Support [NVIDIA Fast-dLLM-v2](https://huggingface.co/Efficient-Large-Model/Fast_dLLM_v2_7B) with HierarchyBlock algorithm. Based on the AR init for Fast-dllm-V2 here: https://github.com/sgl-project/sglang/pull/15013 (Credit to @baonudesifeizhai), I started a new PR because the parallel decoding involves many changes.  

We would like to make the Fast-dLLM-v2 the standard for integrating new dLLMs and their corresponding algorithms, especially to show:
-  How do we implement the algorithms without aggressively changing the systems?
-  How do we test the models and decoding algorithms?

## Modifications

- Init NVIDIA Fast-dLLM-v2 with full logits.
- Init HierarchyBlock algorithm for parallel generation.
- Add two tests for Fast-dLLM-v2 in GSM8k and HierarchyBlock[to improve the accuracy].
- [Todo] support batching when rebase with the newest version after this PR: https://github.com/sgl-project/sglang/pull/14883 

Next steps:
- Update HF implementation of Fast-dLLM-v2 with standard ModelOutput including logits and hidden for unit tests. @Hills-code
- Figure out the logits differences that come from in SGLang Fast-dLLM-v2, compared with the HF implementation.
- Support batching for the HierarchyBlock algorithm. Token shifts will be challenging in batching because we need to map token from the last block with the dLLM requests in a batch, without breaking the schedule batch abstraction. CC @ClawSeven 
## Accuracy Tests

GSM8k test in test/srt/dllm/test_fast_dllm_v2.py:
- Qwen-2.5-7B-Instruct: 86.5%
- Fast-dLLM-v2+AR implementation: 74.5%
- Fast-dLLM-v2+HierarchyBlock: 49.5%(To fix)

I post failure cases here: https://paste.ubuntu.com/p/cQf8GDb9dB/ 
Even the first generated token could be different with the prompt: 

> Explain how Low Confidence decoding works.

To run the test, check: test/srt/dllm/test_hierarchy_block.py 

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: test/srt/dllm/test_hierarchy_block.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            "hidden_states": hidden_states,
+            "prompt_ids": input_ids_list,
+            "formatted_prompt": formatted_prompt,
+        }
+
+    def _get_sglang_outputs(self, prompt: str, max_new_tokens: int):
+        formatted_prompt = self._format_prompt_with_chat_template(prompt)
+        prompt_ids = self.tokenizer(formatted_prompt, return_tensors="pt")["input_ids"][0].tolist()
+        
+        # FIXME(Jinwei): Add return_logprob and return_hidden_states support
```
[the remark this answers is not in the document]
Comment (Monstertail) [reply test/srt/dllm/test_hierarchy_block.py:143]: @hills-code Could you please add standard ModelOutput in Huggingface Fast-dLLM-v2 modeling?
We need logits and hidden to compare sglang and HF implementation: https://huggingface.co/Efficient-Large-Model/Fast_dLLM_v2_7B/blob/main/modeling.py .

File: test/srt/dllm/test_hierarchy_block.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+    def _get_hf_outputs(self, prompt: str, max_new_tokens: int):
+        formatted_prompt = self._format_prompt_with_chat_template(prompt)
+        inputs = self.tokenizer(formatted_prompt, return_tensors="pt").to(self.device)
+        prompt_token_count = inputs["input_ids"].shape[1]
+        input_ids_list = inputs["input_ids"][0].tolist()
+        
+        with torch.no_grad():
+            # FIXME(Jinwei): HF Fast dLLM v2 doesn't generate full max_new_tokens, but max_new_tokens-prompt_length tokens.
+            outputs = self.hf_model.generate(
```
[the remark this answers is not in the document]
Comment (Monstertail) [reply test/srt/dllm/test_hierarchy_block.py:103]: Also, HF Fast dLLM v2 doesn't generate full max_new_tokens, but max_new_tokens-prompt_length tokens.
Therefore, if we set `max_new_tokens=64`,  in SGLang it may generate 64 tokens, while in HF it generates `64-prompt_len tokens`. @hills-code
